beacon-chain/db/kv/slasher.go

```diff
@@ -0,0 +1,175 @@
+	err := s.db.View(func(tx *bolt.Tx) error {
+		bkt := tx.Bucket(epochByValidatorBucket)
+		enc, err := validatorIdx.MarshalSSZ()
+		if err != nil {
+			return err
+		}
+		epochBytes := bkt.Get(enc)
+		if epochBytes == nil {
+			return nil
+		}
+		exists = true
+		return epoch.UnmarshalSSZ(epochBytes)
+	})
+	return epoch, exists, err
+}
+
+// SaveLatestEpochAttestedForValidators updates the latest epoch a slice
+// of validator indices has attested to.
+func (s *Store) SaveLatestEpochAttestedForValidators(
+	ctx context.Context, validatorIndices []types.ValidatorIndex, epoch types.Epoch,
+) error {
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.SaveLatestEpochAttestedForValidator")
+	defer span.End()
+	return s.db.Update(func(tx *bolt.Tx) error {
+		bkt := tx.Bucket(epochByValidatorBucket)
+		for _, valIdx := range validatorIndices {
+			key, err := valIdx.MarshalSSZ()
+			if err != nil {
+				return err
+			}
+			val, err := epoch.MarshalSSZ()
+			if err != nil {
+				return err
+			}
+			if err := bkt.Put(key, val); err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+// AttestationRecordForValidator given a validator index and a target epoch,
+// retrieves an existing attestation record we have stored in the database.
+func (s *Store) AttestationRecordForValidator(
+	ctx context.Context, validatorIdx types.ValidatorIndex, targetEpoch types.Epoch,
+) (*slashertypes.AttestationRecord, error) {
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.AttestationRecordForValidator")
+	defer span.End()
+	var record *slashertypes.AttestationRecord
+	err := s.db.View(func(tx *bolt.Tx) error {
+		bkt := tx.Bucket(attestationRecordsBucket)
+		encIdx, err := validatorIdx.MarshalSSZ()
+		if err != nil {
+			return err
+		}
+		encEpoch, err := targetEpoch.MarshalSSZ()
+		if err != nil {
+			return err
+		}
+		key := append(encIdx, encEpoch...)
+		value := bkt.Get(key)
+		if value == nil {
+			return nil
+		}
+		var sr [32]byte
+		copy(sr[:], value[16:])
+		record = &slashertypes.AttestationRecord{
+			Source:      ssz.UnmarshallUint64(value[0:8]),
+			Target:      ssz.UnmarshallUint64(value[8:16]),
+			SigningRoot: sr,
+		}
+		return nil
+	})
+	return record, err
+}
+
+// SaveAttestationRecordForValidator saves an attestation record for a validator index.
+func (s *Store) SaveAttestationRecordForValidator(
+	ctx context.Context,
+	validatorIdx types.ValidatorIndex,
+	signingRoot [32]byte,
+	attestation *ethpb.IndexedAttestation,
+) error {
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.SaveAttestationRecordForValidator")
+	defer span.End()
+	return s.db.Update(func(tx *bolt.Tx) error {
+		bkt := tx.Bucket(attestationRecordsBucket)
+		encIdx, err := validatorIdx.MarshalSSZ()
+		if err != nil {
+			return err
+		}
+		encEpoch := ssz.MarshalUint64(make([]byte, 0), attestation.Data.Target.Epoch)
+		key := append(encIdx, encEpoch...)
+		value := make([]byte, 48)
+		copy(value[0:8], ssz.MarshalUint64(make([]byte, 0), attestation.Data.Source.Epoch))
```

Comment:
I think encode `IndexedAttestation` to `value` should be its own helper, and vice versa for decode